# Major Map: An AI-Powered Academic Planner and Predictor for SJSU Students

## 1. Data Loading

In [1]:
!rm -rf major_map
!git clone https://github.com/7HE-LUCKY-FISH/major_map.git

Cloning into 'major_map'...
remote: Enumerating objects: 571, done.
remote: Counting objects: 100% (311/311), done.
remote: Compressing objects: 100% (250/250), done.
remote: Total 571 (delta 120), reused 172 (delta 53), pack-reused 260 (from 1)
Receiving objects: 100% (571/571), 1.31 MiB | 16.14 MiB/s, done.
Resolving deltas: 100% (188/188), done.


In [2]:
!ls /content/

encoders.pkl  major_map  sample_data


In [3]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import LabelEncoder
from pathlib import Path

In [4]:
# Path to repo in Colab
data_dir = Path("major_map/data/csv_data")

# Get all csv files
csv_files = sorted(data_dir.glob("*.csv"))

print("Found CSV files:")
for f in csv_files:
    print(" -", f.name)

# Combine all the csv files into one
# Check method number 4: https://medium.com/@stella96joshua/how-to-combine-multiple-csv-files-using-python-for-your-analysis-a88017c6ff9e
df_original = pd.concat([pd.read_csv(f) for f in csv_files], ignore_index=True)

Found CSV files:
 - Fall-2022.csv
 - Fall-2023.csv
 - Fall-2024.csv
 - Spring-2022.csv
 - Spring-2023.csv
 - Spring-2024.csv
 - Spring-2025.csv


## 2. Data Understanding

### 2a. Basic Inspection

In [5]:
df_original.head()

Section  Number          Mode              Title   Satifies  \
0  BIOL 10 (Section 01)   40529     In Person   The Living World     GE: B2   
1  BIOL 10 (Section 03)   40060     In Person   The Living World     GE: B2   
2  BIOL 10 (Section 04)   47603  Fully Online   The Living World     GE: B2   
3  BIOL 10 (Section 99)   41828  Fully Online   The Living World     GE: B2   
4  CHEM 1A (Section 01)   40081     In Person  General Chemistry  GE: B1+B3   

   Unit Type Days            Times       Instructor Location  \
0     3  LEC   TR  09:00AM-10:15AM  Allison Harness   SCI164   
1     3  LEC   MW  10:30AM-11:45AM  Phillip Hawkins   SCI164   
2     3  LEC  TBA              TBA  Phillip Hawkins   ONLINE   
3     3  LEC  TBA              TBA  Mary Poffenroth   ONLINE   
4     5  LEC  MWF  09:30AM-10:20AM       Resa Kelly   SCI142   

               Dates  Seats  Year Semester  
0  08/19/22-12/06/22     59  2022     Fall  
1  08/19/22-12/06/22     42  2022     Fall  
2  08/19/22-12/06/22      6  2022     Fall  
3  08/19/22-12/06/22      1  2022     Fall  
4  08/19/22-12/06/22      0  2022     Fall

In [6]:
df_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4007 entries, 0 to 4006
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Section     4007 non-null   object
 1   Number      4007 non-null   int64 
 2   Mode        4007 non-null   object
 3   Title       4007 non-null   object
 4   Satifies    1615 non-null   object
 5   Unit        4007 non-null   int64 
 6   Type        4007 non-null   object
 7   Days        4007 non-null   object
 8   Times       4007 non-null   object
 9   Instructor  4007 non-null   object
 10  Location    3945 non-null   object
 11  Dates       4007 non-null   object
 12  Seats       4007 non-null   int64 
 13  Year        4007 non-null   int64 
 14  Semester    4007 non-null   object
dtypes: int64(4), object(11)
memory usage: 469.7+ KB


In [7]:
df_original['Year'].value_counts()

Year
2024    1191
2022    1127
2023    1115
2025     574
Name: count, dtype: int64

In [8]:
df_original['Semester'].value_counts()

Semester
Spring    2160
Fall      1847
Name: count, dtype: int64

In [9]:
df_original['Instructor'].nunique()

646

In [10]:
df_original['Section'].nunique()

980

### 2b. Check Missing / Special Values

In [11]:
df_original['Times'].value_counts().head(10)

Times
10:30AM-11:45AM    371
12:00PM-01:15PM    366
09:00AM-10:15AM    316
01:30PM-02:45PM    300
03:00PM-04:15PM    263
04:30PM-05:45PM    185
06:00PM-08:45PM    138
TBA                102
09:00AM-11:45AM    102
03:00PM-05:45PM     97
Name: count, dtype: int64

In [12]:
df_original['Days'].value_counts().head(10)

Days
MW      1120
TR       989
F        400
T        343
W        310
R        289
M        229
MTWR     113
TBA      102
MWF       86
Name: count, dtype: int64

In [13]:
print(df_original['Instructor'].value_counts())
print('\n-----------------------------\n')
print(df_original['Section'].value_counts())
print('\n-----------------------------\n')
print(df_original['Semester'].value_counts())

Instructor
Richard Low                                 53
Padmavati Tanniru                           52
Alla Petrosyan                              51
Olga Kovaleva                               48
Medha Bodas                                 47
                                            ..
Neomi Millan                                 1
Peter Beyersdorf / Kenneth Wharton           1
Ehsan Khatami                                1
Nargis Adham / Azadeh Shahid Faylienejad     1
Resa Kelly                                   1
Name: count, Length: 646, dtype: int64

-----------------------------

Section
ENGR 100W (Section 16)    14
ENGR 100W (Section 14)    14
ENGR 100W (Section 06)    14
ENGR 100W (Section 18)    11
ENGR 100W (Section 12)    10
                          ..
CMPE 165 (Section 80)      1
CS 100W (Section 84)       1
CS 146 (Section 81)        1
CS 146 (Section 82)        1
CS 147 (Section 81)        1
Name: count, Length: 980, dtype: int64

-----------------------------

Sem

## 3. Data Preprocessing

In this section we will go through each features and analyze what is the best way to preprocess the data so they can be useful to use for the machine leanring model.

In [14]:
df_preprocess = df_original.copy()
df_preprocess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4007 entries, 0 to 4006
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Section     4007 non-null   object
 1   Number      4007 non-null   int64 
 2   Mode        4007 non-null   object
 3   Title       4007 non-null   object
 4   Satifies    1615 non-null   object
 5   Unit        4007 non-null   int64 
 6   Type        4007 non-null   object
 7   Days        4007 non-null   object
 8   Times       4007 non-null   object
 9   Instructor  4007 non-null   object
 10  Location    3945 non-null   object
 11  Dates       4007 non-null   object
 12  Seats       4007 non-null   int64 
 13  Year        4007 non-null   int64 
 14  Semester    4007 non-null   object
dtypes: int64(4), object(11)
memory usage: 469.7+ KB


### 3a. "Section" Feature

Ex: BIOL 10 (Section 01)
* Just in case, convert every value in "Section" to a string type. If something was not a string (like a number or a missing value NaN), it becomes a string (NaN --> "nan").

* Removes any leading and trailing whitespaces from each string. This is so we can split "Section" into "Dept" and "CourseNumber", or just "Course".

In [15]:
df_preprocess['Section'] = df_preprocess['Section'].astype(str).str.strip()
print("Missing Section:", df_preprocess['Section'].isna().sum())
print(df_preprocess['Section'].head())

Missing Section: 0
0    BIOL 10 (Section 01)
1    BIOL 10 (Section 03)
2    BIOL 10 (Section 04)
3    BIOL 10 (Section 99)
4    CHEM 1A (Section 01)
Name: Section, dtype: object


### 3b. "Number" Feature

Ex: 40529
* Preprocessing "Number" by ensure it's integer, and have no missing.

* "Number" feature might not be even use because this feature is not really meaningful or important.

In [16]:
df_preprocess['Number'] = df_preprocess['Number'].astype(int)
print("Missing Number:", df_preprocess['Number'].isna().sum())
print(df_preprocess['Number'].head())


Missing Number: 0
0    40529
1    40060
2    47603
3    41828
4    40081
Name: Number, dtype: int64


### 3c. "Mode" Feature

Ex: "In Person", "Fully Online", "Hybird"

* There are only 3 unique possible values for mode. We will one-hot encoding it later.

In [17]:
print("Unique Mode values:", df_preprocess['Mode'].unique())

Unique Mode values: ['In Person' 'Fully Online' 'Hybrid']


### 3d. "Title" Feature

Ex: "In Person", "Fully Online", "Hybird"

* Also not really important feature, but just clean it just in case.

In [18]:
df_preprocess['Title'] = df_preprocess['Title'].astype(str).str.strip()
print("Missing Title:", df_preprocess['Title'].isna().sum())
print(df_preprocess['Title'].head())

Missing Title: 0
0     The Living World
1     The Living World
2     The Living World
3     The Living World
4    General Chemistry
Name: Title, dtype: object


### 3e. "Satifies" Feature

* Can be use to track if it satifies for GE area or not. Fill missing (NaN) which MajorOnly.

In [19]:
print("Unique Satifies values:", df_preprocess['Satifies'].unique())
df_preprocess['Satifies'] = df_preprocess['Satifies'].fillna('MajorOnly')
print('------------------------------------------------------------')
print("Missing Satifies:", df_preprocess['Satifies'].isna().sum())

Unique Satifies values: ['GE: B2' 'GE: B1+B3' nan 'GE: S' 'GE: V' 'GE: WID' 'GE: A2' 'GE: C2'
 'GE: E' 'GE: WID+R' 'GE: B4' 'GE: 5B' 'GE: 5A+5C' 'GE: 4' 'GE: 3'
 'GE: 1A' 'GE: 3B' 'GE: WID+3' 'GE: 2']
------------------------------------------------------------
Missing Satifies: 0


### 3f. "Unit" Feature

* The unit can tell us about the hours and type of the course/section.

In [20]:
print(df_preprocess['Unit'].value_counts().sort_index())

Unit
0     905
1     234
2      60
3    2405
4     384
5      19
Name: count, dtype: int64


### 3g. "Type" Feature

* The types are LEC, SEM and LAB. Will do one hot encoding later.

In [21]:
print("Unique Type values:", df_preprocess['Type'].unique())

Unique Type values: ['LEC' 'SEM' 'LAB']


### 3h. "Days" Feature

* Give us the day patterns. TBA is a special flag. TBA is usually online and asyn classes.

In [22]:
print("Missing Days:", df_preprocess['Days'].isna().sum())
print("Unique Days values:", df_preprocess['Days'].unique())
print(df_preprocess['Days'].value_counts().sort_index())

Missing Days: 0
Unique Days values: ['TR' 'MW' 'TBA' 'MWF' 'F' 'T' 'W' 'R' 'M' 'S' 'MTWR']
Days
F        400
M        229
MTWR     113
MW      1120
MWF       86
R        289
S         26
T        343
TBA      102
TR       989
W        310
Name: count, dtype: int64


### 3i. "Times" Feature

* Give the time range (start time and end time) of a section

In [23]:
print("Missing Times:", df_preprocess['Times'].isna().sum())
print(df_preprocess['Times'].value_counts())

Missing Times: 0
Times
10:30AM-11:45AM    371
12:00PM-01:15PM    366
09:00AM-10:15AM    316
01:30PM-02:45PM    300
03:00PM-04:15PM    263
                  ... 
01:30PM-02:24PM      1
03:00PM-03:45PM      1
12:00AM-01:15AM      1
06:00PM-08:20PM      1
08:00AM-09:15AM      1
Name: count, Length: 114, dtype: int64


### 3j. "Instructor" Feature

* Give us the professor name

In [24]:
print("Missing Instructor:", df_preprocess['Instructor'].isna().sum())
print("Number of unique instructors:", df_preprocess['Instructor'].nunique())
print(df_preprocess['Instructor'].unique)

Missing Instructor: 0
Number of unique instructors: 646
<bound method Series.unique of 0                 Allison Harness
1                 Phillip Hawkins
2                 Phillip Hawkins
3                 Mary Poffenroth
4                      Resa Kelly
                  ...            
4002    Azadeh Shahid Faylienejad
4003      Vakini Santhanakrishnan
4004               Ramen Bahuguna
4005                 Nargis Adham
4006               Ramen Bahuguna
Name: Instructor, Length: 4007, dtype: object>


### 3k. "Location" Feature

* Give us the room the lecutre is held in. Online class location is lablelled as "Online"

In [25]:
print("Missing Location before fill:", df_preprocess['Location'].isna().sum())
df_preprocess['Location'] = df_preprocess['Location'].fillna('Unknown')
print("Missing Location after fill:", df_preprocess['Location'].isna().sum())
print('---------------------------------------------------------------')
print(df_preprocess['Location'].value_counts())
print('---------------------------------------------------------------')
print(df_preprocess['Location'].unique())

Missing Location before fill: 62
Missing Location after fill: 0
---------------------------------------------------------------
Location
ONLINE    366
MH424     143
MH224     105
MH323     101
ENG392     97
         ... 
BBC203      1
DMH161      1
BBC126      1
BBC107      1
ENG336      1
Name: count, Length: 195, dtype: int64
---------------------------------------------------------------
['SCI164' 'ONLINE' 'SCI142' 'MD101' 'DH412' 'DH506' 'DH507' 'ENG325'
 'ENG405' 'ENG337' 'ENG489' 'ENG343' 'ENG341' 'ENG286' 'ENG301' 'ENG331'
 'BBC003' 'ENG288' 'ENG206' 'CL222' 'DMH234' 'MH323' 'MH424' 'MH523'
 'MH222' 'MH223' 'SCI311' 'WSQ109' 'BBC202' 'MH225' 'MH422' 'MH233'
 'CL243' 'SCI258' 'DH450' 'DH351' 'BBC004' 'Unknown' 'ENG189' 'ENG258'
 'ENG290' 'ENG305' 'ENG345' 'ENG307' 'ENG238' 'ENG317' 'ENG319' 'ENG321'
 'ENG244' 'ENG289' 'ENG291' 'ENG376' 'BBC124' 'SH411' 'BBC121' 'BBC128'
 'BBC122' 'SH348' 'DMH354' 'BBC123' 'BBC221' 'CL316' 'BBC130' 'CL225B'
 'SH444' 'CL225A' 'BBC225' 'DMH347' 'SH4

### 3l. "Dates" Feature

* Not really important, we already have Year and Semester

In [26]:
print("Missing Dates:", df_preprocess['Dates'].isna().sum())
print(df_preprocess['Dates'].value_counts())

Missing Dates: 0
Dates
08/21/24-12/09/24    631
08/21/23-12/06/23    626
08/19/22-12/06/22    590
01/23/25-05/12/25    574
01/24/24-05/13/24    560
01/26/22-05/16/22    537
01/25/23-05/15/23    489
Name: count, dtype: int64


### 3m. "Seats" Feature

* Number of seats left for a section. Not really useful, will drop this feature.

In [27]:
df_preprocess['Seats'] = df_preprocess['Seats'].astype(int)
print("Missing Seats:", df_preprocess['Seats'].isna().sum())
print(df_preprocess['Seats'].describe())

Missing Seats: 0
count    4007.000000
mean        3.279760
std         9.260627
min         0.000000
25%         0.000000
50%         0.000000
75%         2.000000
max       181.000000
Name: Seats, dtype: float64


### 3n. "Year" Feature

* Calendar Year: should be only 2022, 2023, 2024 and 2025

In [28]:
print("Missing Year:", df_preprocess['Year'].isna().sum())
print("Years:", df_preprocess['Year'].unique())

Missing Year: 0
Years: [2022 2023 2024 2025]


### 3o. "Semester" Feature

* Term name: We only focus on Fall and Spring (no Winter or Summer)

In [29]:
print("Missing Semester:", df_preprocess['Semester'].isna().sum())
print("Semesters:", df_preprocess['Semester'].unique())

Missing Semester: 0
Semesters: ['Fall' 'Spring']


In [30]:
df_preprocess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4007 entries, 0 to 4006
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Section     4007 non-null   object
 1   Number      4007 non-null   int64 
 2   Mode        4007 non-null   object
 3   Title       4007 non-null   object
 4   Satifies    4007 non-null   object
 5   Unit        4007 non-null   int64 
 6   Type        4007 non-null   object
 7   Days        4007 non-null   object
 8   Times       4007 non-null   object
 9   Instructor  4007 non-null   object
 10  Location    4007 non-null   object
 11  Dates       4007 non-null   object
 12  Seats       4007 non-null   int64 
 13  Year        4007 non-null   int64 
 14  Semester    4007 non-null   object
dtypes: int64(4), object(11)
memory usage: 469.7+ KB


## 4. Data Engineering

In [31]:
# Make a copy for data engineering
df_engineer = df_preprocess.copy()

Originally, we have a total of 15 features:
1. Section          (original)
2. Number           (original, not used as feature)
3. Mode             (original, categorical feature)
4. Title            (original, usually not used as feature in v1)
5. Satifies         (original, optional categorical feature)
6. Unit             (original, numeric feature)
7. Type             (original, categorical feature)
8. Days             (original, used to build slot, not as input feature)
9. Times            (original, used to build StartMinutes/slot)
10. Instructor      (original, used to create instructor_id target)
11. Location        (original)
12. Dates           (original, not used as feature)
13. Seats           (original)
14. Year            (original, numeric feature)
15. Semester        (original, categorical feature)

New Features we will get from data engineering:
1. Dept            (engineered from Section)
2. CourseNumber    (engineered from Section)
3. CourseCode     (engineered from Dept + CourseNumber, optional)
4. HasGE           (engineered from Satifies, 0/1 flag, optional)
5. StartMinutes    (engineered from Times)
6. EndMinutes      (engineered from Times)
7. DurationMinutes (engineered from End-Start)
7. Slot            (engineered from Days + StartMinutes, e.g. "MWF_540")
8. instructor_id   (engineered target from Instructor)
9. slot_id         (engineered target from slot)
10. Building        (engineered from Location, e.g. "SCI", "MD", "ONLINE")
12. Term            (engineered from Year + Semester, e.g. "2022_Fall")
13. SemesterIndex   (engineered from term, 0,1,2,... in time order)

---






### 4a. Section → Dept, CourseNumber, and CourseCode

In [32]:
section = df_engineer['Section'].str.extract(r'^(\w+)\s+([^\s]+)')

df_engineer['Dept'] = section[0]
df_engineer['CourseNumber'] = section[1]
df_engineer['CourseCode'] = df_engineer['Dept'].astype(str) + ' ' + df_engineer['CourseNumber'].astype(str)
df_engineer[['Section', 'Dept', 'CourseNumber', 'CourseCode']].head()

Section  Dept CourseNumber CourseCode
0  BIOL 10 (Section 01)  BIOL           10    BIOL 10
1  BIOL 10 (Section 03)  BIOL           10    BIOL 10
2  BIOL 10 (Section 04)  BIOL           10    BIOL 10
3  BIOL 10 (Section 99)  BIOL           10    BIOL 10
4  CHEM 1A (Section 01)  CHEM           1A    CHEM 1A

### 4b. Times → StartMinutes and EndMinutes

* Minutes after midnight (00:00 am)

In [33]:
def parse_time_range(s):
  if s == "TBA" or "-" not in s:
    return -1, -1, -1

  # s looks like "09:00AM-10:15AM"
  # split the start and end with '-'
  start_str, end_str = s.split('-')

  # parse each part as time
  # format='%I:%M%p'
  # %I = hour in 12-hour clock (01-12)
  # %M = minutes (00-59)
  # %p = AM/PM
  start_dt = pd.to_datetime(start_str, format='%I:%M%p')
  end_dt   = pd.to_datetime(end_str,   format='%I:%M%p')

  # convert to "minutes since midnight"
  start_min = start_dt.hour * 60 + start_dt.minute
  end_min = end_dt.hour * 60 + end_dt.minute

  # section duration
  duration_min = end_min - start_min

  return start_min, end_min, duration_min

# Apply to df_model['Times'] and create three new columns
df_engineer[['StartMinutes', 'EndMinutes', 'DurationMinutes']] = df_engineer['Times'].apply(
    lambda s: pd.Series(parse_time_range(s))
)

df_engineer[['Times', 'StartMinutes', 'EndMinutes', 'DurationMinutes']].head()


Times  StartMinutes  EndMinutes  DurationMinutes
0  09:00AM-10:15AM           540         615               75
1  10:30AM-11:45AM           630         705               75
2              TBA            -1          -1               -1
3              TBA            -1          -1               -1
4  09:30AM-10:20AM           570         620               50

### 4c. Days + StartMinutes + EndMinutes → Slot

* This show use the day and start minute. This feature is important for schedule planning later.

* We will later turn them into Slot_ID

In [34]:
def make_slot(row):
    days = str(row['Days']).strip()
    start = int(row['StartMinutes'])
    end   = int(row['EndMinutes'])

    # If this was a TBA row that you encoded as -1
    if start == -1 or end == -1:
        return days + '_TBA'

    return f"{days}_{start}_{end}"

df_engineer['Slot'] = df_engineer.apply(make_slot, axis=1)
df_engineer[['Days', 'Times', 'StartMinutes', 'Slot']].head()

Days            Times  StartMinutes         Slot
0   TR  09:00AM-10:15AM           540   TR_540_615
1   MW  10:30AM-11:45AM           630   MW_630_705
2  TBA              TBA            -1      TBA_TBA
3  TBA              TBA            -1      TBA_TBA
4  MWF  09:30AM-10:20AM           570  MWF_570_620

### 4d. Satifies → HasGE

* Pretty straightforward. Class that have satifies mean it is in the major requirements. This feature can be good to create roadmaps as well as schedules later.


In [35]:
print(df_engineer['Satifies'].unique())
print('---------------------------------------------')
# 1 if it's a GE area ("GE: ..."), 0 if MajorOnly or anything else
df_engineer['HasGE'] = df_engineer['Satifies'].astype(str).str.startswith('GE:').astype(int)
print('---------------------------------------------')
print(df_engineer[['Satifies', 'HasGE']].head())
print('---------------------------------------------')
print(df_engineer['HasGE'].value_counts())

['GE: B2' 'GE: B1+B3' 'MajorOnly' 'GE: S' 'GE: V' 'GE: WID' 'GE: A2'
 'GE: C2' 'GE: E' 'GE: WID+R' 'GE: B4' 'GE: 5B' 'GE: 5A+5C' 'GE: 4'
 'GE: 3' 'GE: 1A' 'GE: 3B' 'GE: WID+3' 'GE: 2']
---------------------------------------------
---------------------------------------------
    Satifies  HasGE
0     GE: B2      1
1     GE: B2      1
2     GE: B2      1
3     GE: B2      1
4  GE: B1+B3      1
---------------------------------------------
HasGE
0    2392
1    1615
Name: count, dtype: int64


### 4e. Location → Building

* Ex: SCI164 into SCI, and Online stays Online
* Note: MD is Morris Dailey Auditorium, the SJSU website had changed the name into Town Hall


In [36]:
def get_building(location):
  location = str(location).strip()
  if location in ['ONLINE', 'Unknown']:
    return location

  prefix = ''
  for ch in location:
    if ch.isalpha():
      prefix += ch
    else:
      break
  return prefix if prefix else 'Unknown'

df_engineer['Building'] = df_engineer['Location'].apply(get_building)
print(df_engineer['Building'].value_counts())

Building
ENG        1159
MH          662
DH          516
ONLINE      366
SCI         314
BBC         281
SH          188
CL          176
WSQ         138
DMH          90
Unknown      62
MD           24
ISB          15
YUH           9
HGH           2
CCB           2
IS            2
DBH           1
Name: count, dtype: int64


### 4e. Year + Semester → Term

In [37]:
df_engineer['Term'] = df_engineer['Year'].astype(str) + '_' + df_engineer['Semester'].astype(str)
df_engineer[['Year', 'Semester', 'Term']].head()

Year Semester       Term
0  2022     Fall  2022_Fall
1  2022     Fall  2022_Fall
2  2022     Fall  2022_Fall
3  2022     Fall  2022_Fall
4  2022     Fall  2022_Fall

### 4f. Semester → SemesterIndex

* Sort the unique terms and assing 0, 1, 2,... in time order

In [38]:
sem_order = {'Spring': 0, 'Fall': 1}

# Unique (Year, Semester) combos
term_df = (
    df_engineer[['Year', 'Semester']]
    .drop_duplicates()
    .copy()
)


# Sort by Year, then Spring/Fall
term_df['sem_order'] = term_df['Semester'].map(sem_order)
term_df = term_df.sort_values(['Year', 'sem_order']).reset_index(drop=True)

# Assign 0,1,2,... as SemesterIndex
term_df['SemesterIndex'] = term_df.index

print(term_df)

# Merge into df_engineer
df_engineer = df_engineer.merge(
    term_df[['Year', 'Semester', 'SemesterIndex']],
    on=['Year', 'Semester'],
    how='left'
)

print(
    df_engineer[['Year', 'Semester', 'SemesterIndex']]
    .drop_duplicates()
    .sort_values('SemesterIndex')
)


   Year Semester  sem_order  SemesterIndex
0  2022   Spring          0              0
1  2022     Fall          1              1
2  2023   Spring          0              2
3  2023     Fall          1              3
4  2024   Spring          0              4
5  2024     Fall          1              5
6  2025   Spring          0              6
      Year Semester  SemesterIndex
1847  2022   Spring              0
0     2022     Fall              1
2384  2023   Spring              2
590   2023     Fall              3
2873  2024   Spring              4
1216  2024     Fall              5
3433  2025   Spring              6


### 4g. Label Encoding Targets: Instructor, Slot and CourseCode
* Encode each unique instructor, slot, and course code as an integer ID using `LabelEncoder`.

In [39]:
instr_le = LabelEncoder()
df_engineer['Instructor_ID'] = instr_le.fit_transform(df_engineer['Instructor'])

print("Number of unique instructors:", df_engineer['Instructor_ID'].nunique())
df_engineer[['Instructor', 'Instructor_ID']].head()

Number of unique instructors: 646


Instructor  Instructor_ID
0  Allison Harness             35
1  Phillip Hawkins            439
2  Phillip Hawkins            439
3  Mary Poffenroth            363
4       Resa Kelly            471

In [40]:
# LabelEncode Slot → slot_id
slot_le = LabelEncoder()
df_engineer['Slot_ID'] = slot_le.fit_transform(df_engineer['Slot'])

print("Number of unique slots:", df_engineer['Slot_ID'].nunique())
df_engineer[['Slot', 'Slot_ID']].head()

Number of unique slots: 261


Slot  Slot_ID
0   TR_540_615      180
1   MW_630_705       90
2      TBA_TBA      168
3      TBA_TBA      168
4  MWF_570_620       63

In [41]:
course_le = LabelEncoder()
df_engineer['CourseCode_ID'] = course_le.fit_transform(df_engineer['CourseCode'])

print("Number of unique course codes:", df_engineer['CourseCode_ID'].nunique())
df_engineer[['CourseCode', 'CourseCode_ID']].head()

Number of unique course codes: 92


CourseCode  CourseCode_ID
0    BIOL 10              0
1    BIOL 10              0
2    BIOL 10              0
3    BIOL 10              0
4    CHEM 1A              1

In [42]:
print(df_engineer.info())
print("-------------------------------------------")
df_engineer.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4007 entries, 0 to 4006
Data columns (total 29 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Section          4007 non-null   object
 1   Number           4007 non-null   int64 
 2   Mode             4007 non-null   object
 3   Title            4007 non-null   object
 4   Satifies         4007 non-null   object
 5   Unit             4007 non-null   int64 
 6   Type             4007 non-null   object
 7   Days             4007 non-null   object
 8   Times            4007 non-null   object
 9   Instructor       4007 non-null   object
 10  Location         4007 non-null   object
 11  Dates            4007 non-null   object
 12  Seats            4007 non-null   int64 
 13  Year             4007 non-null   int64 
 14  Semester         4007 non-null   object
 15  Dept             4007 non-null   object
 16  CourseNumber     4007 non-null   object
 17  CourseCode       4007 non-null   

Section  Number          Mode              Title   Satifies  \
0  BIOL 10 (Section 01)   40529     In Person   The Living World     GE: B2   
1  BIOL 10 (Section 03)   40060     In Person   The Living World     GE: B2   
2  BIOL 10 (Section 04)   47603  Fully Online   The Living World     GE: B2   
3  BIOL 10 (Section 99)   41828  Fully Online   The Living World     GE: B2   
4  CHEM 1A (Section 01)   40081     In Person  General Chemistry  GE: B1+B3   

   Unit Type Days            Times       Instructor  ... EndMinutes  \
0     3  LEC   TR  09:00AM-10:15AM  Allison Harness  ...        615   
1     3  LEC   MW  10:30AM-11:45AM  Phillip Hawkins  ...        705   
2     3  LEC  TBA              TBA  Phillip Hawkins  ...         -1   
3     3  LEC  TBA              TBA  Mary Poffenroth  ...         -1   
4     5  LEC  MWF  09:30AM-10:20AM       Resa Kelly  ...        620   

  DurationMinutes         Slot  HasGE Building       Term SemesterIndex  \
0              75   TR_540_615      1      SCI  2022_Fall             1   
1              75   MW_630_705      1      SCI  2022_Fall             1   
2              -1      TBA_TBA      1   ONLINE  2022_Fall             1   
3              -1      TBA_TBA      1   ONLINE  2022_Fall             1   
4              50  MWF_570_620      1      SCI  2022_Fall             1   

  Instructor_ID  Slot_ID  CourseCode_ID  
0            35      180              0  
1           439       90              0  
2           439      168              0  
3           363      168              0  
4           471       63              1  

[5 rows x 29 columns]

# Major Map: Neural Network Models

In [43]:
# Neural network dataframe
df_nn = df_engineer.copy()

# Features columns that will be use for neural network
required = ['Dept', 'CourseCode', 'SemesterIndex', 'Instructor', 'Slot', 'Mode', 'Type', 'Building', 'DurationMinutes', 'Unit', 'HasGE', 'StartMinutes']

# IsTBA colum: 1 if start time is unknown, otherwise 0
df_nn['IsTBA'] = (df_nn['StartMinutes'] == -1).astype(int)
print(df_nn['IsTBA'])

# Converts start minutes into half-hour blocks
df_nn['StartBin'] = (df_nn['StartMinutes'] // 30).astype(int)

# Make duration safe if missing
df_nn['DurationMinutes'] = df_nn['DurationMinutes'].fillna(0).astype(int)

# Convert duration into 15-minute buckets
df_nn['DurationBin'] = (df_nn['DurationMinutes'] // 15).astype(int)

0       0
1       0
2       1
3       1
4       0
       ..
4002    0
4003    0
4004    0
4005    0
4006    0
Name: IsTBA, Length: 4007, dtype: int64


In [44]:
import pickle

# columns to encode
categorical_cols = {
    'Dept': 'Dept_ID',
    'CourseCode': 'CourseCode_ID',
    'Mode': 'Mode_ID',
    'Type': 'Type_ID',
    'Building': 'Building_ID',
    'Days': 'Days_ID',
    'Semester': 'Semester_ID',
}

encoders = {}
for col, out_col in categorical_cols.items():
    le = LabelEncoder()
    df_nn[out_col] = le.fit_transform(df_nn[col].astype(str))
    encoders[col] = le

# Targets (Instructor, Slot, CourseCode) - keep course code encoder already fit
instr_le = LabelEncoder()
df_nn['Instructor_ID'] = instr_le.fit_transform(df_nn['Instructor'].astype(str))
encoders['Instructor'] = instr_le

slot_le = LabelEncoder()
df_nn['Slot_ID'] = slot_le.fit_transform(df_nn['Slot'].astype(str))
encoders['Slot'] = slot_le

# Save encoders for inference
with open('encoders.pkl','wb') as f:
    pickle.dump(encoders, f)

In [45]:
# choose sem thresholds according to your data
# Example: train <= sem 5, val == 6, test >= 7 (adjust to your SemesterIndex range)
max_sem = df_nn['SemesterIndex'].max()
# pick sensible cut points; you can set T0, T1 manually
T0 = df_nn['SemesterIndex'].quantile(0.6).astype(int)
T1 = df_nn['SemesterIndex'].quantile(0.8).astype(int)

train_df = df_nn[df_nn['SemesterIndex'] <= T0].reset_index(drop=True)
val_df   = df_nn[(df_nn['SemesterIndex'] > T0) & (df_nn['SemesterIndex'] <= T1)].reset_index(drop=True)
test_df  = df_nn[df_nn['SemesterIndex'] > T1].reset_index(drop=True)

print("Train/Val/Test sizes:", len(train_df), len(val_df), len(test_df))

Train/Val/Test sizes: 2802 631 574
